src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -953,6 +963,90 @@ await using (var enumerator = M().GetAsyncEnumerator())
+        {
+            string source = @"
+using static System.Console;
+class C
+{
+    static async System.Collections.Generic.IAsyncEnumerable<int> M(int value)
+    {
+        Write(""1 "");
+        await System.Threading.Tasks.Task.CompletedTask;
+        Write(""2 "");
+        yield return 3;
+        Write(""4 "");
+        value++;
+        Write($""{value} "");
+    }
+    static async System.Threading.Tasks.Task Main()
+    {
+        var enumerable = M(1);
+        await using (var enumerator1 = enumerable.GetAsyncEnumerator())
+        {
+            await using (var enumerator2 = enumerable.GetAsyncEnumerator())
+            {
+                if (!await enumerator1.MoveNextAsync()) throw null;
+                Write($""Stream1:{enumerator1.Current} "");
+                if (!await enumerator2.MoveNextAsync()) throw null;
+                Write($""Stream2:{enumerator2.Current} "");
+                if (await enumerator1.MoveNextAsync()) throw null;
+                if (await enumerator2.MoveNextAsync()) throw null;
+                Write(""Done"");
+            }
+        }
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_common }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2 Stream1:3 1 2 Stream2:3 4 2 4 2 Done");
+        }
+
+        [ConditionalFact(typeof(WindowsDesktopOnly))]
+        [WorkItem(30275, "https://github.com/dotnet/roslyn/issues/30275")]
+        public void CallingGetEnumeratorTwice2()
+        {
+            string source = @"
+using static System.Console;
+class C
+{
+    static async System.Collections.Generic.IAsyncEnumerable<int> M(int value)
+    {
+        Write(""1 "");
+        await System.Threading.Tasks.Task.CompletedTask;
+        Write(""2 "");
+        yield return 3;
+        Write(""4 "");
+        value++;
+        Write($""{value} "");
+    }
+    static async System.Threading.Tasks.Task Main()
+    {
+        var enumerable = M(1);
+        await using (var enumerator1 = enumerable.GetAsyncEnumerator())
+        {
+            await using (var enumerator2 = enumerable.GetAsyncEnumerator())
+            {
+                if (!await enumerator1.MoveNextAsync()) throw null;
+                Write($""Stream1:{enumerator1.Current} "");
+                if (await enumerator1.MoveNextAsync()) throw null;
+
+                if (!await enumerator2.MoveNextAsync()) throw null;
+                Write($""Stream2:{enumerator2.Current} "");
+                if (await enumerator2.MoveNextAsync()) throw null;
+
+                Write(""Done"");
+            }
+        }
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_common }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2 Stream1:3 4 2 1 2 Stream2:3 4 2 Done");
+        }
```

Comment:
> } [](start = 7, length = 2)

Perhaps test overlapping enumerators with calls to each outside the lifetime of the other.
```
var enumerator1 = enumerable.GetAsyncEnumerator();
await enumerator1.MoveNext();
var enumerator2 = enumerable.GetAsyncEnumerator();
await enumerator2.MoveNext();
await enumerator1.DisposeAsync();
await enumerator2.MoveNext();
await enumerator2.DisposeAsync();
```

Reply:
Good idea. I'll add

Reply:
Agreed, especially with an await in between.